<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->


  <!--radix_placeholder_meta_tags-->
  <title>Practical Bayes Part II</title>

  <meta property="description" itemprop="description" content="Practical steps you can take in your Stan journey to deal with issues and explore your models fully."/>


  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2020-10-30"/>
  <meta property="article:created" itemprop="dateCreated" content="2020-10-30"/>
  <meta name="article:author" content="Michael Clark"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="Practical Bayes Part II"/>
  <meta property="og:type" content="article"/>
  <meta property="og:description" content="Practical steps you can take in your Stan journey to deal with issues and explore your models fully."/>
  <meta property="og:locale" content="en_US"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="Practical Bayes Part II"/>
  <meta property="twitter:description" content="Practical steps you can take in your Stan journey to deal with issues and explore your models fully."/>

  <!--/radix_placeholder_meta_tags-->
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","author","date","preview","output","draft","tags","categories"]}},"value":[{"type":"character","attributes":{},"value":["Practical Bayes Part II"]},{"type":"character","attributes":{},"value":["Practical steps you can take in your Stan journey to deal with issues and explore your models fully.\n"]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","url"]}},"value":[{"type":"character","attributes":{},"value":["Michael Clark"]},{"type":"character","attributes":{},"value":["https://m-clark.github.io"]}]}]},{"type":"character","attributes":{},"value":["October 30, 2020"]},{"type":"character","attributes":{},"value":["../../img/r_and_stan.png"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc","highlight","css"]}},"value":[{"type":"logical","attributes":{},"value":[false]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["pygments"]},{"type":"character","attributes":{},"value":["../../styles.css","https://use.fontawesome.com/releases/v5.14.0/css/all.css"]}]}]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["tags","taggy"]},{"type":"character","attributes":{},"value":["bayesian"]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["how-to-bayes-part-ii_files/anchor-4.2.2/anchor.min.js","how-to-bayes-part-ii_files/bowser-1.9.3/bowser.min.js","how-to-bayes-part-ii_files/crosstalk-1.1.0.1/css/crosstalk.css","how-to-bayes-part-ii_files/crosstalk-1.1.0.1/js/crosstalk.js","how-to-bayes-part-ii_files/crosstalk-1.1.0.1/js/crosstalk.js.map","how-to-bayes-part-ii_files/crosstalk-1.1.0.1/js/crosstalk.min.js","how-to-bayes-part-ii_files/crosstalk-1.1.0.1/js/crosstalk.min.js.map","how-to-bayes-part-ii_files/datatables-binding-0.15/datatables.js","how-to-bayes-part-ii_files/datatables-binding-0.16/datatables.js","how-to-bayes-part-ii_files/datatables-css-0.0.0/datatables-crosstalk.css","how-to-bayes-part-ii_files/distill-2.2.21/template.v2.js","how-to-bayes-part-ii_files/dt-core-1.10.20/css/jquery.dataTables.extra.css","how-to-bayes-part-ii_files/dt-core-1.10.20/css/jquery.dataTables.min.css","how-to-bayes-part-ii_files/dt-core-1.10.20/js/jquery.dataTables.min.js","how-to-bayes-part-ii_files/figure-html5/conditional-effects-model-comparison-1.svg","how-to-bayes-part-ii_files/figure-html5/conditional-effects-model-comparison-2.svg","how-to-bayes-part-ii_files/figure-html5/conditional-effects-model-comparison-3.svg","how-to-bayes-part-ii_files/figure-html5/conditional-effects-model-comparison-4.svg","how-to-bayes-part-ii_files/figure-html5/conditional-effects-model-comparison-5.svg","how-to-bayes-part-ii_files/figure-html5/conditional-effects-model-comparison-show-1.svg","how-to-bayes-part-ii_files/figure-html5/model-baseline-ppcheck-1.svg","how-to-bayes-part-ii_files/figure-html5/model-baseline-ppcheck-2.svg","how-to-bayes-part-ii_files/figure-html5/model-baseline-ppcheck-3.svg","how-to-bayes-part-ii_files/figure-html5/model-baseline-ppcheck-4.svg","how-to-bayes-part-ii_files/figure-html5/model-baseline-ppcheck-5.svg","how-to-bayes-part-ii_files/figure-html5/model-baseline-ppcheck-6.svg","how-to-bayes-part-ii_files/figure-html5/model-baseline-ppcheck-med-max-1.svg","how-to-bayes-part-ii_files/figure-html5/model-baseline-ppcheck-med-max-2.svg","how-to-bayes-part-ii_files/figure-html5/model-baseline-ppcheck-re-1.svg","how-to-bayes-part-ii_files/figure-html5/model-baseline-summary-plots-1.svg","how-to-bayes-part-ii_files/figure-html5/model-baseline-summary-plots-2.svg","how-to-bayes-part-ii_files/figure-html5/model-start-explore-1.svg","how-to-bayes-part-ii_files/figure-html5/ppcheck-1.svg","how-to-bayes-part-ii_files/figure-html5/ppcheck-2.svg","how-to-bayes-part-ii_files/figure-html5/ppcheck-3.svg","how-to-bayes-part-ii_files/figure-html5/ppcheck-4.svg","how-to-bayes-part-ii_files/figure-html5/proposed-priors-plot-1.svg","how-to-bayes-part-ii_files/figure-html5/r2-not-pp-check-1.svg","how-to-bayes-part-ii_files/figure-html5/tidybayesdemo-1.svg","how-to-bayes-part-ii_files/figure-html5/unnamed-chunk-1-1.svg","how-to-bayes-part-ii_files/figure-html5/unnamed-chunk-1-2.svg","how-to-bayes-part-ii_files/figure-html5/unnamed-chunk-1-3.svg","how-to-bayes-part-ii_files/figure-html5/unnamed-chunk-1-4.svg","how-to-bayes-part-ii_files/figure-html5/unnamed-chunk-1-5.svg","how-to-bayes-part-ii_files/figure-html5/unnamed-chunk-4-1.svg","how-to-bayes-part-ii_files/figure-html5/unnamed-chunk-4-2.svg","how-to-bayes-part-ii_files/figure-html5/unnamed-chunk-4-3.svg","how-to-bayes-part-ii_files/figure-html5/unnamed-chunk-4-4.svg","how-to-bayes-part-ii_files/header-attrs-2.5/header-attrs.js","how-to-bayes-part-ii_files/htmlwidgets-1.5.1/htmlwidgets.js","how-to-bayes-part-ii_files/htmlwidgets-1.5.2/htmlwidgets.js","how-to-bayes-part-ii_files/jquery-1.11.3/jquery.min.js","how-to-bayes-part-ii_files/jquery-1.12.4/jquery.min.js","how-to-bayes-part-ii_files/jquery-1.12.4/LICENSE.txt","how-to-bayes-part-ii_files/kePrint-0.0.1/kePrint.js","how-to-bayes-part-ii_files/lightable-0.0.1/lightable.css","how-to-bayes-part-ii_files/webcomponents-2.0.0/webcomponents.js"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: hidden;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure img {
    width: 100%;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme
    $('code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace("index.html", "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="how-to-bayes-part-ii_files/header-attrs-2.5/header-attrs.js"></script>
  <script src="how-to-bayes-part-ii_files/htmlwidgets-1.5.2/htmlwidgets.js"></script>
  <script src="how-to-bayes-part-ii_files/jquery-1.12.4/jquery.min.js"></script>
  <link href="how-to-bayes-part-ii_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
  <script src="how-to-bayes-part-ii_files/datatables-binding-0.16/datatables.js"></script>
  <link href="how-to-bayes-part-ii_files/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet" />
  <link href="how-to-bayes-part-ii_files/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet" />
  <script src="how-to-bayes-part-ii_files/dt-core-1.10.20/js/jquery.dataTables.min.js"></script>
  <link href="how-to-bayes-part-ii_files/crosstalk-1.1.0.1/css/crosstalk.css" rel="stylesheet" />
  <script src="how-to-bayes-part-ii_files/crosstalk-1.1.0.1/js/crosstalk.min.js"></script>
  <script src="how-to-bayes-part-ii_files/kePrint-0.0.1/kePrint.js"></script>
  <link href="how-to-bayes-part-ii_files/lightable-0.0.1/lightable.css" rel="stylesheet" />
  <script src="how-to-bayes-part-ii_files/anchor-4.2.2/anchor.min.js"></script>
  <script src="how-to-bayes-part-ii_files/bowser-1.9.3/bowser.min.js"></script>
  <script src="how-to-bayes-part-ii_files/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="how-to-bayes-part-ii_files/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->

  <link rel="stylesheet" href="../../styles.css" type="text/css"/>
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.14.0/css/all.css" type="text/css"/>

</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"Practical Bayes Part II","description":"Practical steps you can take in your Stan journey to deal with issues and explore your models fully.","authors":[{"author":"Michael Clark","authorURL":"https://m-clark.github.io","affiliation":"&nbsp;","affiliationURL":"#","orcidID":""}],"publishedDate":"2020-10-30T00:00:00.000-04:00","citationText":"Clark, 2020"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>Practical Bayes Part II</h1>
<!--radix_placeholder_categories-->
<div class="dt-tags">
<div class="dt=tag">bayesian</div>
</div>
<!--/radix_placeholder_categories-->
<p><p>Practical steps you can take in your Stan journey to deal with issues and explore your models fully.</p></p>
</div>

<div class="d-byline">
  Michael Clark <a href="https://m-clark.github.io" class="uri">https://m-clark.github.io</a> 
  
<br/>October 30, 2020
</div>

<div class="d-article">
<div class="d-contents d-contents-float">
<nav class="l-text toc figcaption" id="TOC">
<h3>Contents</h3>
<ul>
<li><a href="#better-bayesian-approaches-part-ii">Better Bayesian Approaches, Part II</a>
<ul>
<li><a href="#outline-for-better-bayesian-analysis">Outline for Better Bayesian Analysis</a></li>
</ul></li>
<li><a href="#example-data">Example data</a></li>
<li><a href="#simulate-from-priors">Simulate from priors</a></li>
<li><a href="#run-and-summarize-the-baseline-model">Run and Summarize the Baseline Model</a>
<ul>
<li><a href="#check-priors">Check priors</a></li>
<li><a href="#explore-and-visualize-effects">Explore and visualize effects</a></li>
<li><a href="#model-effectiveness">Model Effectiveness</a>
<ul>
<li><a href="#posterior-predictive-checks">Posterior predictive checks</a></li>
<li><a href="#bayes-r-squared">Bayes R-squared</a></li>
</ul></li>
</ul></li>
<li><a href="#prediction-model-comparison">Prediction &amp; Model Comparison</a>
<ul>
<li><a href="#basic-prediction">Basic prediction</a></li>
<li><a href="#model-comparison">Model Comparison</a>
<ul>
<li><a href="#problems-at-the-loo">Problems at the loo</a></li>
<li><a href="#solutions-with-model-comparison">Solutions with Model Comparison</a></li>
</ul></li>
<li><a href="#averaging-models">Averaging models</a></li>
<li><a href="#cross-validation">Cross-validation</a></li>
</ul></li>
<li><a href="#resources">Resources</a></li>
</ul>
</nav>
</div>
<h1 id="better-bayesian-approaches-part-ii">Better Bayesian Approaches, Part II</h1>
<p>We’ve talked about the basics one can do to run a Bayesian model, and what to do if there is a problem, which was the primary goal of <a href="">Part I</a>. But it might be nice if we could avoid the problems in the first place, and our model might still be inadequate without any warnings. So let’s engage in some better practices you can use every time to help things run more smoothly, and get more from your models.</p>
<h2 id="outline-for-better-bayesian-analysis">Outline for Better Bayesian Analysis</h2>
<div style="text-align: center">
<p><i class="fas fa-list-ol fa-5x" style="padding: 20px; color: #ff5500"></i></p>
</div>
<p><br></p>
<div style="text-align: center">
<p><i class="fas fa-hat-wizard fa-5x" style="padding: 20px; color: #ff5500"></i></p>
</div>
<p><br></p>
<p>We’ll cover the following steps in more detail, but here is a general outline.</p>
<ul>
<li>First generate ‘fake data’ to assess prior viability</li>
<li>With adequate priors, start with a simple, but plausible model</li>
<li>For simple models you do not need many iterations
<ul>
<li>If you are doing standard GLM or simpler versions of common extensions, even the defaults are likely overkill. For example, a basic linear regression should converge almost immediately.</li>
</ul></li>
<li>Problems (see <a href="">Part I</a>)
<ul>
<li>If Rhat/ESS is issue, run more iterations</li>
<li>If <code>max_treedepth</code> is issue, increase the default</li>
<li>If divergent transitions
<ul>
<li>Check the data, is there some problem that went undiscovered?</li>
<li>Check the data, has it been scaled properly?</li>
<li>Can something more be done about priors?</li>
<li>Use visualizations to aid diagnosis</li>
<li>Reparameterize model (unlikely if using a higher level package)</li>
<li>Get more/better data (unlikely in practice)</li>
<li>Get a better model</li>
</ul></li>
<li>Issues with loo (<a href="#problems-at-the-loo">see below</a>)</li>
</ul></li>
<li>Use posterior predictive checks
<ul>
<li>Nice, but what if it doesn’t fit?
<ul>
<li>Get better data (unlikely in practice)</li>
<li>Get a better model</li>
<li>If nothing to do, advise others how to get better data/models in the future.</li>
</ul></li>
</ul></li>
<li>Explore a more viable model
<ul>
<li>Add interactions</li>
<li>Add nonlinear relations</li>
<li>Account for other structure (e.g. random effects)</li>
</ul></li>
<li>Compare and/or average models</li>
</ul>
<p>We’ll now demonstrate these steps.</p>
<h1 id="example-data">Example data</h1>
<div style="text-align: center">
<p><i class="fas fa-database fa-5x" style="padding: 20px; color: #ff5500"></i></p>
</div>
<p><br></p>
<p>NEED LINK</p>
<p><a href="">As before</a>, I’m going to create some data for us to run some basic models with, the same as Part I. As a reminder, the true underlying model has categorical and continuous covariates, interactions, nonlinear relationships, random effects (observations are clustered in groups), and some variables are collinear.</p>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<p>For our purposes so we’ll create a data frame with the total sample size of 1000.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># create the primary data frame</span>

<span class='va'>main_df</span> <span class='op'>=</span> 
  <span class='fu'>create_data</span><span class='op'>(</span>N <span class='op'>=</span> <span class='fl'>1000</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>as_tibble</span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>select</span><span class='op'>(</span><span class='va'>group</span>, <span class='va'>b1</span><span class='op'>:</span><span class='va'>x3</span>, <span class='va'>y</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>mutate</span><span class='op'>(</span>
    b1 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>b1</span><span class='op'>)</span>,   <span class='co'># will help with visuals</span>
    b2 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>b2</span><span class='op'>)</span>
  <span class='op'>)</span>
</code></pre>
</div>
</div>
<h1 id="simulate-from-priors">Simulate from priors</h1>
<p>A first step is to produce some viable priors. But the obvious question is, what priors should we choose? Thankfully, for standard models there is not much guesswork involved. Bayesian analysis has been around a long time, so the bulk of the work has been done for you. Even default settings should not affect things much, especially for rstanarm, which has some basic defaults that are informed by the data. However, due to the flexibility of the brms modeling functions, some priors are unspecified and left flat (i.e. uniform), which is something we definitely don’t want. And even defaults could still cause problems for more complex situations. So how might we choose better ones?</p>
<p>The basic idea here is to generate parameters (e.g. regression coefficients) based on the prior distributions for those parameters, predict data based on those prior draws, and then compare the predictions to our observed target variable. The brms package makes this very easy to do. We will check the following types of priors.</p>
<div class="layout-chunk" data-layout="l-body">
<div id="htmlwidget-b00327701f375583695d" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-b00327701f375583695d">{"x":{"filter":"none","data":[["Note:","Regression coefficients:","Intercept:","Variances:"],["similar to brms","uniform","default","default"],["similar to rstanarm","normal, diffuse","default","default"],["similar to rstanarm","normal","default","default"],["If data is standardized, this would be very reasonable","Normal(0, 1)","default","default"],["restricts range of intercept to more plausible values","Normal(0, 1)","mean of `y` (~3)","default"],["restricts range of sigma to more plausible values","Normal(0, 1)","based on mean of `y` (~3)","based on sd of `y` (~1)"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Prior set 0<\/th>\n      <th>Prior set 1<\/th>\n      <th>Prior set 2<\/th>\n      <th>Prior set 3<\/th>\n      <th>Prior set 4<\/th>\n      <th>Prior set 5<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"t","autoWidth":true,"columnDefs":[{"targets":"","orderable":false},{"orderable":false,"targets":0}],"scrollX":true,"order":[],"orderClasses":false,"rowCallback":"function(row, data) {\nvar value=data[-1]; $(this.api().cell(row, -1).node()).css({'white-space':'nowrap'});\nvar value=data[0]; $(this.api().cell(row, 0).node()).css({'white-space':'nowrap'});\nvar value=data[1]; $(this.api().cell(row, 1).node()).css({'white-space':'nowrap'});\nvar value=data[2]; $(this.api().cell(row, 2).node()).css({'white-space':'nowrap'});\nvar value=data[3]; $(this.api().cell(row, 3).node()).css({'white-space':'nowrap'});\nvar value=data[4]; $(this.api().cell(row, 4).node()).css({'white-space':'nowrap'});\nvar value=data[5]; $(this.api().cell(row, 5).node()).css({'white-space':'nowrap'});\nvar value=data[0]; $(this.api().cell(row, 0).node()).css({'font-weight':'bold','text-align':'left'});\n}"}},"evals":["options.rowCallback"],"jsHooks":[]}</script>
</div>
<p>We can use <span class="func" style="">pp_check</span> to examine the prior-generated data versus the observed target <code>y</code>, but I wait to show them all together at the end.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/paul-buerkner/brms'>brms</a></span><span class='op'>)</span>

<span class='va'>pr_uniform</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/brms/man/set_prior.html'>prior</a></span><span class='op'>(</span><span class='fu'>uniform</span><span class='op'>(</span><span class='op'>-</span><span class='fl'>100</span>, <span class='fl'>100</span><span class='op'>)</span>, lb <span class='op'>=</span> <span class='op'>-</span><span class='fl'>100</span>, ub <span class='op'>=</span> <span class='fl'>100</span>, <span class='st'>'b'</span><span class='op'>)</span>

<span class='va'>model_default_prior</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/brms/man/brm.html'>brm</a></span><span class='op'>(</span>
  <span class='va'>y</span> <span class='op'>~</span> <span class='va'>b1</span> <span class='op'>+</span> <span class='va'>b2</span> <span class='op'>+</span> <span class='va'>x1</span> <span class='op'>+</span> <span class='va'>x2</span> <span class='op'>+</span> <span class='va'>x3</span>, 
  data <span class='op'>=</span> <span class='va'>main_df</span>,
  iter <span class='op'>=</span> <span class='fl'>1000</span>,
  sample_prior <span class='op'>=</span> <span class='st'>'only'</span>,
  prior <span class='op'>=</span> <span class='va'>pr_uniform</span>
<span class='op'>)</span>

<span class='co'># pp_check(model_default_prior, nsamples = 50)</span>

<span class='co'># diffuse normal for reg</span>
<span class='va'>pr_norm_b_0_10</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/brms/man/set_prior.html'>prior</a></span><span class='op'>(</span><span class='fu'>normal</span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>10</span><span class='op'>)</span>, <span class='st'>'b'</span><span class='op'>)</span>

<span class='va'>model_0_norm_b_0_10</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/brms/man/brm.html'>brm</a></span><span class='op'>(</span>
  <span class='va'>y</span> <span class='op'>~</span> <span class='va'>b1</span> <span class='op'>+</span> <span class='va'>b2</span> <span class='op'>+</span> <span class='va'>x1</span> <span class='op'>+</span> <span class='va'>x2</span> <span class='op'>+</span> <span class='va'>x3</span>, 
  data <span class='op'>=</span> <span class='va'>main_df</span>,
  iter <span class='op'>=</span> <span class='fl'>1000</span>,
  sample_prior <span class='op'>=</span> <span class='st'>'only'</span>,
  prior <span class='op'>=</span> <span class='va'>pr_norm_b_0_10</span>
<span class='op'>)</span>

<span class='co'># pp_check(model_0_norm_b_0_10, nsamples = 50)</span>

<span class='co'># rstanarm-like prior</span>
<span class='va'>pr_auto</span> <span class='op'>=</span> <span class='fu'>sjstats</span><span class='fu'>::</span><span class='fu'><a href='https://strengejacke.github.io/sjstats/reference/auto_prior.html'>auto_prior</a></span><span class='op'>(</span>
  <span class='va'>y</span> <span class='op'>~</span> <span class='va'>b1</span> <span class='op'>+</span> <span class='va'>b2</span> <span class='op'>+</span> <span class='va'>x1</span> <span class='op'>+</span> <span class='va'>x2</span> <span class='op'>+</span> <span class='va'>x3</span>,
  data <span class='op'>=</span> <span class='va'>main_df</span>,
  gaussian <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span>

<span class='va'>model_auto_prior</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/brms/man/brm.html'>brm</a></span><span class='op'>(</span>
  <span class='va'>y</span> <span class='op'>~</span> <span class='va'>b1</span> <span class='op'>+</span> <span class='va'>b2</span> <span class='op'>+</span> <span class='va'>x1</span> <span class='op'>+</span> <span class='va'>x2</span> <span class='op'>+</span> <span class='va'>x3</span>, 
  data <span class='op'>=</span> <span class='va'>main_df</span>,
  iter <span class='op'>=</span> <span class='fl'>1000</span>,
  sample_prior <span class='op'>=</span> <span class='st'>'only'</span>,
  prior <span class='op'>=</span> <span class='va'>pr_auto</span>
<span class='op'>)</span>

<span class='co'># pp_check(model_auto_prior, nsamples = 50)</span>


<span class='va'>pr_norm_b_0_1</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/brms/man/set_prior.html'>prior</a></span><span class='op'>(</span><span class='fu'>normal</span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>1</span><span class='op'>)</span>, <span class='st'>'b'</span><span class='op'>)</span>

<span class='va'>model_0_norm_b_0_1</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/brms/man/brm.html'>brm</a></span><span class='op'>(</span>
  <span class='va'>y</span> <span class='op'>~</span> <span class='va'>b1</span> <span class='op'>+</span> <span class='va'>b2</span> <span class='op'>+</span> <span class='va'>x1</span> <span class='op'>+</span> <span class='va'>x2</span> <span class='op'>+</span> <span class='va'>x3</span>, 
  data <span class='op'>=</span> <span class='va'>main_df</span>,
  iter <span class='op'>=</span> <span class='fl'>1000</span>,
  sample_prior <span class='op'>=</span> <span class='st'>'only'</span>,
  prior <span class='op'>=</span> <span class='va'>pr_norm_b_0_1</span>
<span class='op'>)</span>

<span class='co'># pp_check(model_0_norm_b_0_1, nsamples = 50)</span>

<span class='va'>pr_norm_b_norm_int</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/pkg/brms/man/set_prior.html'>prior</a></span><span class='op'>(</span><span class='fu'>normal</span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>1</span><span class='op'>)</span>, class <span class='op'>=</span> <span class='st'>'b'</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/pkg/brms/man/set_prior.html'>prior</a></span><span class='op'>(</span><span class='fu'>normal</span><span class='op'>(</span><span class='fl'>3</span>, <span class='fl'>1</span><span class='op'>)</span>, class <span class='op'>=</span> <span class='st'>'Intercept'</span><span class='op'>)</span><span class='co'>#,</span>
<span class='op'>)</span>

<span class='va'>model_0_norm_b_0_1_norm_Int</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/brms/man/brm.html'>brm</a></span><span class='op'>(</span>
  <span class='va'>y</span> <span class='op'>~</span> <span class='va'>b1</span> <span class='op'>+</span> <span class='va'>b2</span> <span class='op'>+</span> <span class='va'>x1</span> <span class='op'>+</span> <span class='va'>x2</span> <span class='op'>+</span> <span class='va'>x3</span>, 
  data <span class='op'>=</span> <span class='va'>main_df</span>,
  iter <span class='op'>=</span> <span class='fl'>1000</span>,
  sample_prior <span class='op'>=</span> <span class='st'>'only'</span>,
  prior <span class='op'>=</span> <span class='va'>pr_norm_b_norm_int</span>
<span class='op'>)</span>

<span class='co'># pp_check(model_0_norm_b_0_1_norm_Int, nsamples = 50)</span>


<span class='va'>pr_norm_b_norm_int_t_sigma</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/pkg/brms/man/set_prior.html'>prior</a></span><span class='op'>(</span><span class='fu'>normal</span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>1</span><span class='op'>)</span>, class <span class='op'>=</span> <span class='st'>'b'</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/pkg/brms/man/set_prior.html'>prior</a></span><span class='op'>(</span><span class='fu'>normal</span><span class='op'>(</span><span class='fl'>3</span>, <span class='fl'>1</span><span class='op'>)</span>, class <span class='op'>=</span> <span class='st'>'Intercept'</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/pkg/brms/man/set_prior.html'>prior</a></span><span class='op'>(</span><span class='fu'>student_t</span><span class='op'>(</span><span class='fl'>10</span>, <span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span>, class <span class='op'>=</span> <span class='st'>'sigma'</span><span class='op'>)</span>
<span class='op'>)</span>

<span class='va'>model_0_norm_b_0_1_norm_Int_sigma</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/brms/man/brm.html'>brm</a></span><span class='op'>(</span>
  <span class='va'>y</span> <span class='op'>~</span> <span class='va'>b1</span> <span class='op'>+</span> <span class='va'>b2</span> <span class='op'>+</span> <span class='va'>x1</span> <span class='op'>+</span> <span class='va'>x2</span> <span class='op'>+</span> <span class='va'>x3</span>, 
  data <span class='op'>=</span> <span class='va'>main_df</span>,
  iter <span class='op'>=</span> <span class='fl'>1000</span>,
  sample_prior <span class='op'>=</span> <span class='st'>'only'</span>,
  prior <span class='op'>=</span> <span class='va'>pr_norm_b_norm_int_t_sigma</span>
<span class='op'>)</span>

<span class='co'># pp_check(model_0_norm_b_0_1_norm_Int_sigma, nsamples = 50)</span>
</code></pre>
</div>
</div>
<p>The following plot shows the model predictions based on priors only. We restrict the range of values for display purposes, so note that some of the priors would generate more extreme results. For example, the default prior setting could generate values into the <span class="math inline">\(\pm\)</span> 500 and beyond. I also mark the boundaries of the observed target variable.</p>
<div class="layout-chunk" data-layout="l-body">
<p><img src="how-to-bayes-part-ii_files/figure-html5/proposed-priors-plot-1.svg" width="624" style="display: block; margin: auto;" /></p>
</div>
<p>So given that our target variable is between -2 and 8, it seems that just adding some basic, data-informed information to our priors resulted in more plausible results. This will generally help our models be more efficient and better behaved. Note that if all else fails, you can use a convenience function like <span class="func" style="">auto_prior</span> demonstrated above.</p>
<h1 id="run-and-summarize-the-baseline-model">Run and Summarize the Baseline Model</h1>
<p>Now let’s run a baseline model, one that’s simple but plausible. Given that there will eventually be additional complexities, I’ll go ahead and add some iterations, and increase max tree depth and adapt delta now to make the code reusable.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/paul-buerkner/brms'>brms</a></span><span class='op'>)</span>

<span class='va'>pr</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/pkg/brms/man/set_prior.html'>prior</a></span><span class='op'>(</span><span class='fu'>normal</span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>1</span><span class='op'>)</span>, class <span class='op'>=</span> <span class='st'>'b'</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/pkg/brms/man/set_prior.html'>prior</a></span><span class='op'>(</span><span class='fu'>student_t</span><span class='op'>(</span><span class='fl'>10</span>, <span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span>, class <span class='op'>=</span> <span class='st'>'sigma'</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/pkg/brms/man/set_prior.html'>prior</a></span><span class='op'>(</span><span class='fu'>student_t</span><span class='op'>(</span><span class='fl'>10</span>, <span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span>, class <span class='op'>=</span> <span class='st'>'sd'</span><span class='op'>)</span>  <span class='co'># prior for random intercept std dev</span>
<span class='op'>)</span>

<span class='va'>model_baseline</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/brms/man/brm.html'>brm</a></span><span class='op'>(</span>
  <span class='va'>y</span> <span class='op'>~</span> <span class='va'>b1</span> <span class='op'>+</span> <span class='va'>b2</span> <span class='op'>+</span> <span class='va'>x1</span> <span class='op'>+</span> <span class='va'>x2</span> <span class='op'>+</span> <span class='va'>x3</span> <span class='op'>+</span> <span class='op'>(</span><span class='fl'>1</span> <span class='op'>|</span> <span class='va'>group</span><span class='op'>)</span>, 
  data    <span class='op'>=</span> <span class='va'>main_df</span>,
  warmup  <span class='op'>=</span> <span class='fl'>5000</span>,
  iter    <span class='op'>=</span> <span class='fl'>6000</span>,
  thin    <span class='op'>=</span> <span class='fl'>4</span>,
  prior   <span class='op'>=</span> <span class='va'>pr</span>, 
  cores   <span class='op'>=</span> <span class='fl'>4</span>,
  control <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
    adapt_delta   <span class='op'>=</span> <span class='fl'>.95</span>,
    max_treedepth <span class='op'>=</span> <span class='fl'>15</span>
  <span class='op'>)</span>,
  save_pars <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/brms/man/save_pars.html'>save_pars</a></span><span class='op'>(</span>all <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>  <span class='co'># potentially allows for more  more post-processing functionality</span>
<span class='op'>)</span>
</code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/r/base/summary.html'>summary</a></span><span class='op'>(</span><span class='va'>model_baseline</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code> Family: gaussian 
  Links: mu = identity; sigma = identity 
Formula: y ~ b1 + b2 + x1 + x2 + x3 + (1 | group) 
   Data: main_df (Number of observations: 1000) 
Samples: 4 chains, each with iter = 6000; warmup = 5000; thin = 4;
         total post-warmup samples = 1000

Group-Level Effects: 
~group (Number of levels: 100) 
              Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
sd(Intercept)     0.86      0.07     0.73     1.02 1.00      699      830

Population-Level Effects: 
          Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
Intercept     2.69      0.21     2.27     3.08 1.01      859      881
b11           0.80      0.07     0.67     0.93 1.00      962      920
b21          -0.14      0.18    -0.50     0.22 1.01     1009      964
x1            0.03      0.04    -0.04     0.11 1.00      963      914
x2           -0.03      0.04    -0.10     0.04 1.00     1077      949
x3            0.27      0.04     0.20     0.34 1.00      830      845

Family Specific Parameters: 
      Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
sigma     1.03      0.02     0.99     1.08 1.00      877      846

Samples were drawn using sampling(NUTS). For each parameter, Bulk_ESS
and Tail_ESS are effective sample size measures, and Rhat is the potential
scale reduction factor on split chains (at convergence, Rhat = 1).</code></pre>
</div>
<p>For reporting purposes all you need are the <code>Estimate</code> and lower and upper bounds. If you want a visual approach, you can use something like the following types of plots.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/brms/man/mcmc_plot.brmsfit.html'>mcmc_plot</a></span><span class='op'>(</span><span class='va'>model_baseline</span>, type <span class='op'>=</span> <span class='st'>'areas'</span><span class='op'>)</span>
</code></pre>
</div>
<img src="how-to-bayes-part-ii_files/figure-html5/model-baseline-summary-plots-1.svg" width="624" style="display: block; margin: auto;" />
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/brms/man/mcmc_plot.brmsfit.html'>mcmc_plot</a></span><span class='op'>(</span><span class='va'>model_baseline</span>, type <span class='op'>=</span> <span class='st'>'intervals'</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="how-to-bayes-part-ii_files/figure-html5/model-baseline-summary-plots-2.svg" width="624" style="display: block; margin: auto;" /></p>
</div>
<p>The tidybayes package offers some nice options as well.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://mjskay.github.io/tidybayes'>tidybayes</a></span><span class='op'>)</span>

<span class='co'># get_variables(model_baseline) %&gt;% as_tibble() # to see variable names as required for plotting.</span>

<span class='co'># grab fixed effects - intercept</span>
<span class='va'>tidy_plot_data_fe</span> <span class='op'>=</span> <span class='va'>model_baseline</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='http://mjskay.github.io/tidybayes/reference/spread_draws.html'>spread_draws</a></span><span class='op'>(</span><span class='va'>`^b_(b|x).*`</span>, regex <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>pivot_longer</span><span class='op'>(</span><span class='va'>b_b11</span><span class='op'>:</span><span class='va'>b_x3</span>, names_to <span class='op'>=</span> <span class='st'>'coefficient'</span><span class='op'>)</span>

<span class='va'>tidy_plot_data_fe</span> <span class='op'>%&gt;%</span>
  <span class='fu'>ggplot</span><span class='op'>(</span><span class='fu'>aes</span><span class='op'>(</span>y <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/rev.html'>rev</a></span><span class='op'>(</span><span class='va'>coefficient</span><span class='op'>)</span>, x <span class='op'>=</span> <span class='va'>value</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>geom_vline</span><span class='op'>(</span>xintercept <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>.25</span>, <span class='fl'>.25</span><span class='op'>)</span>, color <span class='op'>=</span> <span class='st'>'gray92'</span>, size <span class='op'>=</span> <span class='fl'>.5</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='http://mjskay.github.io/ggdist/reference/geom_dotsinterval.html'>stat_dotsinterval</a></span><span class='op'>(</span>
    <span class='fu'>aes</span><span class='op'>(</span>fill <span class='op'>=</span> <span class='fu'>stat</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/MathFun.html'>abs</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='op'>&lt;</span> <span class='fl'>.25</span><span class='op'>)</span><span class='op'>)</span>,
    quantiles <span class='op'>=</span> <span class='fl'>40</span>,
    point_color <span class='op'>=</span> <span class='st'>'#b2001d'</span>,
    interval_color <span class='op'>=</span> <span class='st'>'#b2001d'</span>,
    interval_alpha <span class='op'>=</span> <span class='fl'>.6</span>
  <span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>scico</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/scico/man/ggplot2-scales.html'>scale_fill_scico_d</a></span><span class='op'>(</span>begin <span class='op'>=</span> <span class='fl'>.2</span>, end <span class='op'>=</span> <span class='fl'>.6</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>labs</span><span class='op'>(</span>y <span class='op'>=</span> <span class='st'>''</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>guides</span><span class='op'>(</span>fill <span class='op'>=</span> <span class='st'>'none'</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>theme_clean</span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="how-to-bayes-part-ii_files/figure-html5/tidybayesdemo-1.svg" width="624" style="display: block; margin: auto;" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<h2 id="check-priors">Check priors</h2>
<p>We might still be concerned about how influential our priors were. So how can we check whether our priors were informative? The following will do a simple check of whether the posterior standard deviation is greater than 10% of the prior standard deviation<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>. Having an informative prior isn’t really a problem in my opinion, unless it’s more informative than you wanted. For example, shrinkage of a coefficient towards zero will generally help avoid overfitting.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/brms/man/prior_summary.brmsfit.html'>prior_summary</a></span><span class='op'>(</span><span class='va'>model_baseline</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>                  prior     class      coef group resp dpar nlpar bound       source
           normal(0, 1)         b                                               user
           normal(0, 1)         b       b11                             (vectorized)
           normal(0, 1)         b       b21                             (vectorized)
           normal(0, 1)         b        x1                             (vectorized)
           normal(0, 1)         b        x2                             (vectorized)
           normal(0, 1)         b        x3                             (vectorized)
 student_t(3, 2.9, 2.5) Intercept                                            default
    student_t(10, 1, 1)        sd                                               user
    student_t(10, 1, 1)        sd           group                       (vectorized)
    student_t(10, 1, 1)        sd Intercept group                       (vectorized)
    student_t(10, 1, 1)     sigma                                               user</code></pre>
<div class="sourceCode">
<pre><code><span class='fu'>bayestestR</span><span class='fu'>::</span><span class='fu'><a href='https://easystats.github.io/bayestestR/reference/check_prior.html'>check_prior</a></span><span class='op'>(</span><span class='va'>model_baseline</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>    Parameter Prior_Quality
1 b_Intercept   informative
2       b_b11 uninformative
3       b_b21   informative
4        b_x1 uninformative
5        b_x2 uninformative
6        b_x3 uninformative</code></pre>
</div>
<p>These results suggest that we might be more informative, but for the intercept, which we largely care less about, and for the factor that is highly unbalanced, which has no obvious solution. I would be fine with this result.</p>
<h2 id="explore-and-visualize-effects">Explore and visualize effects</h2>
<p>We can plot effects easily with brms. The <span class="func" style="">conditional_effects</span> function is what we want here. Without interactions or other things going, on they aren’t very interesting, but it’s a useful tool nonetheless.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/brms/man/conditional_effects.brmsfit.html'>conditional_effects</a></span><span class='op'>(</span><span class='va'>model_baseline</span>, <span class='st'>'b2'</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="how-to-bayes-part-ii_files/figure-html5/model-start-explore-1.svg" width="624" style="display: block; margin: auto;" /></p>
</div>
<p>We can also use the hypothesis function to test for specific types of effects. By default they provide a one-sided probability and uncertainty interval. For starters, we can just duplicate what we saw in the previous summary for the <code>b2</code> effect. The only benefit is to easily obtain the one-sided p-value (that <code>b2</code> is less than zero) and the corresponding <em>evidence ratio</em>, which is just <code>p/(1-p)</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/brms/man/hypothesis.brmsfit.html'>hypothesis</a></span><span class='op'>(</span><span class='va'>model_baseline</span>, <span class='st'>'b21 &lt; 0'</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>Hypothesis Tests for class b:
  Hypothesis Estimate Est.Error CI.Lower CI.Upper Evid.Ratio Post.Prob Star
1  (b21) &lt; 0    -0.14      0.18    -0.44     0.16       3.42      0.77     
---
&#39;CI&#39;: 90%-CI for one-sided and 95%-CI for two-sided hypotheses.
&#39;*&#39;: For one-sided hypotheses, the posterior probability exceeds 95%;
for two-sided hypotheses, the value tested against lies outside the 95%-CI.
Posterior probabilities of point hypotheses assume equal prior probabilities.</code></pre>
</div>
<p>But we can really try anything, which is the power of this function. The following tests whether the combined effect of our categorical covariates is greater than zero.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/brms/man/hypothesis.brmsfit.html'>hypothesis</a></span><span class='op'>(</span><span class='va'>model_baseline</span>, <span class='st'>'b11 + abs(b21) &gt; 0'</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>Hypothesis Tests for class b:
          Hypothesis Estimate Est.Error CI.Lower CI.Upper Evid.Ratio Post.Prob Star
1 (b11+abs(b21)) &gt; 0     0.99      0.15     0.78     1.25        Inf         1    *
---
&#39;CI&#39;: 90%-CI for one-sided and 95%-CI for two-sided hypotheses.
&#39;*&#39;: For one-sided hypotheses, the posterior probability exceeds 95%;
for two-sided hypotheses, the value tested against lies outside the 95%-CI.
Posterior probabilities of point hypotheses assume equal prior probabilities.</code></pre>
</div>
<h2 id="model-effectiveness">Model Effectiveness</h2>
<p>A natural question to ask is how useful our model actually is, which then suggests we need to know how to define such utility. Such an assessment definitely cannot be made with something like ‘statistical significance’. Science of any kind is nothing without prediction, so we we can start there.</p>
<h3 id="posterior-predictive-checks">Posterior predictive checks</h3>
<p>Posterior predictive checks are a key component of bayesian analysis. The prior checks we did before are just a special case of this. Here we actually use the posterior distributions of parameters to generate the data to compare what the model implies and what we actually observe. Doing so can give insight to where the model fails.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/brms/man/pp_check.brmsfit.html'>pp_check</a></span><span class='op'>(</span><span class='va'>model_baseline</span>, nsamples <span class='op'>=</span> <span class='fl'>100</span><span class='op'>)</span>
</code></pre>
</div>
<img src="how-to-bayes-part-ii_files/figure-html5/model-baseline-ppcheck-1.svg" width="624" style="display: block; margin: auto;" />
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/brms/man/pp_check.brmsfit.html'>pp_check</a></span><span class='op'>(</span><span class='va'>model_baseline</span>, nsamples <span class='op'>=</span> <span class='fl'>10</span>, type <span class='op'>=</span><span class='st'>'error_scatter_avg'</span>, alpha <span class='op'>=</span> <span class='fl'>.1</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="how-to-bayes-part-ii_files/figure-html5/model-baseline-ppcheck-2.svg" width="624" style="display: block; margin: auto;" /></p>
</div>
<p>In this case we see good alignment between model and data, and no obvious pattern to the types of errors we are getting. We do see that the model does not capture the most extreme value well, but that’s not terribly surprising.</p>
<p>Consider predictions with and without random effects. You have to look closely, but the random effect predictions appear to do better with the predicting more in the tails.</p>
<div class="layout-chunk" data-layout="l-body">
<p><img src="how-to-bayes-part-ii_files/figure-html5/model-baseline-ppcheck-re-1.svg" width="624" style="display: block; margin: auto;" /></p>
</div>
<p>We can use the same approach to look at specific statistical measures. The following suggests our model is okay with capturing the minimum value, but typically underestimates the maximum value, which we noted earlier.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/brms/man/pp_check.brmsfit.html'>pp_check</a></span><span class='op'>(</span><span class='va'>model_baseline</span>, nsamples <span class='op'>=</span> <span class='fl'>100</span>, type <span class='op'>=</span><span class='st'>'stat'</span>, stat<span class='op'>=</span><span class='st'>'median'</span><span class='op'>)</span>
</code></pre>
</div>
<img src="how-to-bayes-part-ii_files/figure-html5/model-baseline-ppcheck-med-max-1.svg" width="624" style="display: block; margin: auto;" />
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/brms/man/pp_check.brmsfit.html'>pp_check</a></span><span class='op'>(</span><span class='va'>model_baseline</span>, nsamples <span class='op'>=</span> <span class='fl'>100</span>, type <span class='op'>=</span><span class='st'>'stat'</span>, stat <span class='op'>=</span> <span class='st'>'max'</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="how-to-bayes-part-ii_files/figure-html5/model-baseline-ppcheck-med-max-2.svg" width="624" style="display: block; margin: auto;" /></p>
</div>
<p>We can define any function to use for our posterior predictive check. The following shows how to examine the 10th and 90th quantiles. Minimum and maximum values are unlikely to be captured very well due to their inherent variability, so looking at less extreme quantiles (e.g. 10<sup>th</sup> or 90<sup>th</sup> percentile) might be a better way to assess whether the model captures the tails of a distribution.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='va'>q10</span> <span class='op'>=</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>y</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/stats/quantile.html'>quantile</a></span><span class='op'>(</span><span class='va'>y</span>, <span class='fl'>0.1</span><span class='op'>)</span>
<span class='va'>q90</span> <span class='op'>=</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>y</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/stats/quantile.html'>quantile</a></span><span class='op'>(</span><span class='va'>y</span>, <span class='fl'>0.9</span><span class='op'>)</span>

<span class='fu'><a href='https://rdrr.io/pkg/brms/man/pp_check.brmsfit.html'>pp_check</a></span><span class='op'>(</span><span class='va'>model_baseline</span>, nsamples <span class='op'>=</span> <span class='fl'>100</span>, type <span class='op'>=</span><span class='st'>'stat'</span>, stat <span class='op'>=</span> <span class='st'>'q90'</span><span class='op'>)</span>

<span class='fu'><a href='https://rdrr.io/pkg/brms/man/pp_check.brmsfit.html'>pp_check</a></span><span class='op'>(</span><span class='va'>model_baseline</span>, nsamples <span class='op'>=</span> <span class='fl'>100</span>, type <span class='op'>=</span><span class='st'>'stat_2d'</span>, stat <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>'q10'</span>, <span class='st'>'q90'</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<h3 id="bayes-r-squared">Bayes R-squared</h3>
<p>In this scenario, we can examine the amount of variance accounted for in the target variable by the covariates. I don’t really recommend this beyond linear models assuming a normal distribution for the target, but people like to report it. Conceptually, it is simply a (squared) correlation of fitted values to the observed, so can be seen as descriptive statistic. Since we are Bayesians, we also get a ready-made interval for it, as it is based on the posterior predictive distribution. But to stress the complexity in trying to assess this, in this mixed model we can obtain the result with the random effect included (conditional) or without (unconditional). Both are reasonable ways to express the statistic, but the one including the group effect naturally will be superior, assuming the variance is notable in the first place.</p>
<p><a href="https://avehtari.github.io/bayes_R2/bayes_R2.html" class="uri">https://avehtari.github.io/bayes_R2/bayes_R2.html</a></p>
<p>Andrew Gelman, Ben Goodrich, Jonah Gabry, and Aki Vehtari (2018). R-squared for Bayesian regression models. The American Statistician, <a href="doi:10.1080/00031305.2018.1549100" class="uri">doi:10.1080/00031305.2018.1549100</a>. <a href="http://www.stat.columbia.edu/~gelman/research/unpublished/bayes_R2_v3.pdf">Online Preprint</a>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/brms/man/bayes_R2.brmsfit.html'>bayes_R2</a></span><span class='op'>(</span><span class='va'>model_baseline</span><span class='op'>)</span>                   <span class='co'># random effects included</span>
</code></pre>
</div>
<pre><code>    Estimate  Est.Error      Q2.5     Q97.5
R2 0.4816803 0.01802656 0.4458734 0.5151165</code></pre>
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/brms/man/bayes_R2.brmsfit.html'>bayes_R2</a></span><span class='op'>(</span><span class='va'>model_baseline</span>, re_formula <span class='op'>=</span> <span class='cn'>NA</span><span class='op'>)</span>  <span class='co'># random effects not included</span>
</code></pre>
</div>
<pre><code>    Estimate  Est.Error       Q2.5     Q97.5
R2 0.1166801 0.01422869 0.08934349 0.1455141</code></pre>
<div class="sourceCode">
<pre><code><span class='co'># performance::r2_bayes(model_baseline)    # alternative provides both</span>
</code></pre>
</div>
</div>
<p>To show the limitation of R<sup>2</sup>, I rerun the model using a restrictive prior on the intercept. Intercepts for the resulting models are different but the other fixed effects are basically the same. The R<sup>2</sup> suggests equal fit.</p>
<div class="layout-chunk" data-layout="l-body">
<pre><code>Running /Library/Frameworks/R.framework/Resources/bin/R CMD SHLIB foo.c
clang -mmacosx-version-min=10.13 -I&quot;/Library/Frameworks/R.framework/Resources/include&quot; -DNDEBUG   -I&quot;/Users/micl/Rlibs/Rcpp/include/&quot;  -I&quot;/Users/micl/Rlibs/RcppEigen/include/&quot;  -I&quot;/Users/micl/Rlibs/RcppEigen/include/unsupported&quot;  -I&quot;/Users/micl/Rlibs/BH/include&quot; -I&quot;/Users/micl/Rlibs/StanHeaders/include/src/&quot;  -I&quot;/Users/micl/Rlibs/StanHeaders/include/&quot;  -I&quot;/Users/micl/Rlibs/RcppParallel/include/&quot;  -I&quot;/Users/micl/Rlibs/rstan/include&quot; -DEIGEN_NO_DEBUG  -DBOOST_DISABLE_ASSERTS  -DBOOST_PENDING_INTEGER_LOG2_HPP  -DSTAN_THREADS  -DBOOST_NO_AUTO_PTR  -include &#39;/Users/micl/Rlibs/StanHeaders/include/stan/math/prim/mat/fun/Eigen.hpp&#39;  -D_REENTRANT -DRCPP_PARALLEL_USE_TBB=1   -I/usr/local/include   -fPIC  -Wall -g -O2  -c foo.c -o foo.o
In file included from &lt;built-in&gt;:1:
In file included from /Users/micl/Rlibs/StanHeaders/include/stan/math/prim/mat/fun/Eigen.hpp:13:
In file included from /Users/micl/Rlibs/RcppEigen/include/Eigen/Dense:1:
In file included from /Users/micl/Rlibs/RcppEigen/include/Eigen/Core:88:
/Users/micl/Rlibs/RcppEigen/include/Eigen/src/Core/util/Macros.h:613:1: error: unknown type name &#39;namespace&#39;
namespace Eigen {
^
/Users/micl/Rlibs/RcppEigen/include/Eigen/src/Core/util/Macros.h:613:16: error: expected &#39;;&#39; after top level declarator
namespace Eigen {
               ^
               ;
In file included from &lt;built-in&gt;:1:
In file included from /Users/micl/Rlibs/StanHeaders/include/stan/math/prim/mat/fun/Eigen.hpp:13:
In file included from /Users/micl/Rlibs/RcppEigen/include/Eigen/Dense:1:
/Users/micl/Rlibs/RcppEigen/include/Eigen/Core:96:10: fatal error: &#39;complex&#39; file not found
#include &lt;complex&gt;
         ^~~~~~~~~
3 errors generated.
make: *** [foo.o] Error 1</code></pre>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
model
</th>
<th style="text-align:right;">
Estimate
</th>
<th style="text-align:right;">
Est.Error
</th>
<th style="text-align:right;">
Q2.5
</th>
<th style="text-align:right;">
Q97.5
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
baseline
</td>
<td style="text-align:right;">
0.117
</td>
<td style="text-align:right;">
0.014
</td>
<td style="text-align:right;">
0.089
</td>
<td style="text-align:right;">
0.146
</td>
</tr>
<tr>
<td style="text-align:left;">
modified
</td>
<td style="text-align:right;">
0.116
</td>
<td style="text-align:right;">
0.014
</td>
<td style="text-align:right;">
0.089
</td>
<td style="text-align:right;">
0.144
</td>
</tr>
</tbody>
</table>
</div>
<p>However, a posterior predictive check shows clearly the failure of the modified model to capture the data.</p>
<div class="layout-chunk" data-layout="l-body">
<p><img src="how-to-bayes-part-ii_files/figure-html5/r2-not-pp-check-1.svg" width="624" style="display: block; margin: auto;" /></p>
</div>
<p>A variant of R<sup>2</sup>, the ‘LOO’ R<sup>2</sup>, is also available via the <span class="func" style="">loo_R2</span> function. LOO stands for <em>leave-one-out</em>, as in leave-one-out cross-validation. It’s based on the residuals from the leave one out predictions. You can think of it as a better ‘adjusted R<sup>2</sup>’. The results suggests that the LOO R<sup>2</sup> actually picks up the difference in models, and would be lower for the modified model, even if we included the random effects</p>
<div class="layout-chunk" data-layout="l-body">
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
model
</th>
<th style="text-align:right;">
LOO R2
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
baseline
</td>
<td style="text-align:right;">
0.101
</td>
</tr>
<tr>
<td style="text-align:left;">
modified
</td>
<td style="text-align:right;">
0.096
</td>
</tr>
</tbody>
</table>
</div>
<!-- ### Model comparison -->
<!-- We will use estimates like WAIC and loo for model comparison later.  They are essentially used as you would AIC, or root mean squared error to compare models in a predictive fashion.  In addition We only really have our  -->
<!-- Even now, we can use loo as a diagnostic to possibly discover problematic observations, -->
<!-- ```{r} -->
<!-- loo(model_baseline) -->
<!-- WAIC(model_baseline) -->
<!-- post_prob(model_baseline, model_r2_vs_pp) -->
<!-- ``` -->
<!-- The interesting thing here is that we have a grossly inefficient model, yet none we have nothing that notes any issues -->
<h1 id="prediction-model-comparison">Prediction &amp; Model Comparison</h1>
<p>In general, a model is judged most by whether it has practical value. Even if we think a model is effective, there still might be another model that can do better. So, if we’re doing what we should, we generally will have a couple models to compare with one another. And one of the best ways to compare them is via prediction, especially by predicting on data the model wasn’t trained on to begin with.</p>
<p>For our demonstration, we will add two new models. The first adds interactions, the second adds a nonlinear relationship for one of the variables to that model, and is the closest to the underlying data generating mechanism.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='va'>model_interact</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/update.html'>update</a></span><span class='op'>(</span>
  <span class='va'>model_baseline</span>,
  <span class='va'>.</span> <span class='op'>~</span> <span class='va'>.</span> <span class='op'>+</span> <span class='va'>b1</span><span class='op'>:</span><span class='va'>b2</span> <span class='op'>+</span> <span class='va'>b2</span><span class='op'>:</span><span class='va'>x1</span>,
  cores <span class='op'>=</span> <span class='fl'>4</span>
<span class='op'>)</span>

<span class='va'>model_interact_nonlin</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/update.html'>update</a></span><span class='op'>(</span>
  <span class='va'>model_interact</span>,
  <span class='va'>.</span> <span class='op'>~</span> <span class='va'>.</span> <span class='op'>+</span> <span class='fu'><a href='https://rdrr.io/pkg/brms/man/s.html'>s</a></span><span class='op'>(</span><span class='va'>x3</span><span class='op'>)</span>,
  cores <span class='op'>=</span> <span class='fl'>4</span>
<span class='op'>)</span>
</code></pre>
</div>
</div>
<h2 id="basic-prediction">Basic prediction</h2>
<p>With models in hand, let’s look at our basic predictive capabilities. We can get fitted values which include ‘confidence’ intervals, or predictions, which include ‘prediction’ intervals that include the uncertainty for a new observation. We can specify these as follows. First we create a small data set to make some predictions on. It will include both values for of the binary covariates, and the means of the numeric (0).</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='va'>prediction_data</span> <span class='op'>=</span> <span class='fu'>crossing</span><span class='op'>(</span>
  b1 <span class='op'>=</span> <span class='fl'>0</span><span class='op'>:</span><span class='fl'>1</span>,
  b2 <span class='op'>=</span> <span class='fl'>0</span><span class='op'>:</span><span class='fl'>1</span>,
  x1 <span class='op'>=</span> <span class='fl'>0</span>,
  x2 <span class='op'>=</span> <span class='fl'>0</span>,
  x3 <span class='op'>=</span> <span class='fl'>0</span>
<span class='op'>)</span>

<span class='co'># fitted values</span>
<span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/stats/fitted.values.html'>fitted</a></span><span class='op'>(</span><span class='va'>model_baseline</span><span class='op'>)</span><span class='op'>)</span>  
</code></pre>
</div>
<pre><code>     Estimate Est.Error     Q2.5    Q97.5
[1,] 3.498778 0.3384366 2.840337 4.155334
[2,] 3.776796 0.3394366 3.103155 4.441631
[3,] 3.845675 0.3351470 3.185096 4.508866
[4,] 4.337709 0.3429396 3.627330 4.993732
[5,] 3.159843 0.3380833 2.502989 3.820989
[6,] 3.889759 0.3335991 3.230206 4.541787</code></pre>
<div class="sourceCode">
<pre><code><span class='co'># new predictions</span>
<span class='fu'><a href='https://rdrr.io/r/stats/predict.html'>predict</a></span><span class='op'>(</span><span class='va'>model_baseline</span>, newdata <span class='op'>=</span> <span class='va'>prediction_data</span>, re_formula <span class='op'>=</span> <span class='cn'>NA</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>     Estimate Est.Error      Q2.5    Q97.5
[1,] 2.692793  1.073933 0.5396395 4.723138
[2,] 2.547446  1.027297 0.4782266 4.589374
[3,] 3.467172  1.065637 1.4146903 5.398930
[4,] 3.337754  1.012264 1.3668239 5.327862</code></pre>
</div>
<p>In general, we’d always like to visualize the predictions. We can do so as we did before with the <span class="func" style="">conditional_effects</span> function. For the third plot of the nonlinear effect, I modify the basic conditional effects plot that brms provides for a slightly cleaner visualization.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='fu'>conditional_effects</span><span class='op'>(</span><span class='va'>model_baseline</span>, effects <span class='op'>=</span> <span class='st'>'x2'</span><span class='op'>)</span>
</code></pre>
</div>
<img src="how-to-bayes-part-ii_files/figure-html5/conditional-effects-model-comparison-1.svg" width="624" style="display: block; margin: auto;" />
<div class="sourceCode">
<pre><code><span class='fu'>conditional_effects</span><span class='op'>(</span><span class='va'>model_interact</span>, effects <span class='op'>=</span> <span class='st'>'x1:b2'</span><span class='op'>)</span>
</code></pre>
</div>
<img src="how-to-bayes-part-ii_files/figure-html5/conditional-effects-model-comparison-2.svg" width="624" style="display: block; margin: auto;" />
<div class="sourceCode">
<pre><code><span class='va'>init</span> <span class='op'>=</span> <span class='fu'>conditional_effects</span><span class='op'>(</span><span class='va'>model_interact_nonlin</span>, effects <span class='op'>=</span> <span class='st'>'x3'</span>, spaghetti <span class='op'>=</span> <span class='cn'>T</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body-outset">
<p><img src="how-to-bayes-part-ii_files/figure-html5/conditional-effects-model-comparison-show-1.svg" width="624" style="display: block; margin: auto;" /></p>
</div>
<h2 id="model-comparison">Model Comparison</h2>
<blockquote>
<p>Sometimes a simple model will outperform a more complex model… Nevertheless, I believe that deliberately limiting the complexity of the model is not fruitful when the problem is evidently complex. Instead, if a simple model is found that outperforms some particular complex model, the appropriate response is to define a different complex model that captures whatever aspect of the problem led to the simple model performing well. ~ Radford Neal</p>
</blockquote>
<p>In typical situations it is good to have competing models, attempting to see if improvements can be made in one way or another. In the Bayesian approach, we can use estimates like <em>WAIC</em> and <em>LOOIC</em> for model comparison, much like you would AIC to compare models in traditional frameworks. The values themselves don’t tell us much, but in comparing models, lower means less predictive error for ‘ic’ metrics, which is what we want. Like in the traditional framework, LOOIC is ~ -2*expected log posterior density (<em>ELPD</em>)<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>, and since we’re Bayesian, we even have estimates of uncertainty for these values as well.</p>
<p>With our new models in place, we can now make some comparisons using <span class="func" style="">loo_compare</span>. It shows the ‘best’, i.e. lowest valued, model first followed by the others. The result regards the total expected log probability for the leave-one-out observations. We also get stuff like <code>p_loo</code>, which is the effective number of parameters, and <code>Pareto k values</code>, which we’ll talk about soon.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='va'>model_baseline</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/brms/man/add_criterion.html'>add_criterion</a></span><span class='op'>(</span><span class='va'>model_baseline</span>,  <span class='st'>'loo'</span><span class='op'>)</span>
<span class='va'>model_interact</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/brms/man/add_criterion.html'>add_criterion</a></span><span class='op'>(</span><span class='va'>model_interact</span>, <span class='st'>'loo'</span><span class='op'>)</span>
<span class='va'>model_interact_nonlin</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/brms/man/add_criterion.html'>add_criterion</a></span><span class='op'>(</span><span class='va'>model_interact_nonlin</span>, <span class='st'>'loo'</span><span class='op'>)</span>

<span class='co'># example</span>
<span class='fu'><a href='https://rdrr.io/pkg/brms/man/loo.brmsfit.html'>loo</a></span><span class='op'>(</span><span class='va'>model_baseline</span><span class='op'>)</span> 
</code></pre>
</div>
<pre><code>
Computed from 1000 by 1000 log-likelihood matrix

         Estimate   SE
elpd_loo  -1500.5 21.9
p_loo        88.0  3.8
looic      3001.0 43.9
------
Monte Carlo SE of elpd_loo is 0.4.

Pareto k diagnostic values:
                         Count Pct.    Min. n_eff
(-Inf, 0.5]   (good)     992   99.2%   277       
 (0.5, 0.7]   (ok)         8    0.8%   176       
   (0.7, 1]   (bad)        0    0.0%   &lt;NA&gt;      
   (1, Inf)   (very bad)   0    0.0%   &lt;NA&gt;      

All Pareto k estimates are ok (k &lt; 0.7).
See help(&#39;pareto-k-diagnostic&#39;) for details.</code></pre>
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/brms/man/loo_compare.brmsfit.html'>loo_compare</a></span><span class='op'>(</span>
  <span class='va'>model_baseline</span>, 
  <span class='va'>model_interact</span>,
  <span class='va'>model_interact_nonlin</span>
<span class='op'>)</span>
</code></pre>
</div>
<pre><code>                      elpd_diff se_diff
model_interact_nonlin   0.0       0.0  
model_interact        -45.1       9.4  
model_baseline        -46.5       9.6  </code></pre>
</div>
<p>Let’s compare several metrics available to us. In this particular setting, all are generally in agreement in the rank order of the models. However, there is no meaningful difference between the baseline and interaction models, while information criteria would more clearly prefer the latter.</p>
<div class="layout-chunk" data-layout="l-body">
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
model
</th>
<th style="text-align:right;">
R2
</th>
<th style="text-align:right;">
loo_R2
</th>
<th style="text-align:right;">
WAIC
</th>
<th style="text-align:right;">
LOOIC
</th>
<th style="text-align:right;">
ELPD
</th>
<th style="text-align:right;">
weight
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
baseline
</td>
<td style="text-align:right;">
0.48
</td>
<td style="text-align:right;">
0.43
</td>
<td style="text-align:right;">
2998.83
</td>
<td style="text-align:right;">
3001.00
</td>
<td style="text-align:right;">
-1500.50
</td>
<td style="text-align:right;">
5.909983e-06
</td>
</tr>
<tr>
<td style="text-align:left;">
interact
</td>
<td style="text-align:right;">
0.48
</td>
<td style="text-align:right;">
0.43
</td>
<td style="text-align:right;">
2995.68
</td>
<td style="text-align:right;">
2998.07
</td>
<td style="text-align:right;">
-1499.03
</td>
<td style="text-align:right;">
4.950540e-07
</td>
</tr>
<tr>
<td style="text-align:left;">
interact_nonlin
</td>
<td style="text-align:right;">
0.53
</td>
<td style="text-align:right;">
0.48
</td>
<td style="text-align:right;">
2905.42
</td>
<td style="text-align:right;">
2907.96
</td>
<td style="text-align:right;">
-1453.98
</td>
<td style="text-align:right;">
9.999936e-01
</td>
</tr>
</tbody>
<tfoot>
<tr>
<td style="padding: 0; border:0;" colspan="100%">
<sup>a</sup> Odds are relative to baseline
</td>
</tr>
</tfoot>
</table>
</div>
<p>While we have several metrics, for model comparison, we’d like to stick to the IC type appoach. As far as using WAIC vs. LOOIC, the latter has better diagnostics for noting whether there are potential problems in using it. In practice, it likely wouldn’t differ much. As we noted, LOOIC reflects the ELPD, which is used in constructing the model weights<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a>, the model weights can then be used in making final predictions.</p>
<h3 id="problems-at-the-loo">Problems at the loo</h3>
<p><a href="https://mc-stan.org/loo/articles/loo2-weights.html" class="uri">https://mc-stan.org/loo/articles/loo2-weights.html</a></p>
<p>issues- overfitting, time, complexity is more accurate portrayal of nature even if not ‘best’</p>
<p>After the model warnings discussed in <a href="">Part I</a>, the next most common point of confusion I see with clients is with model comparison. Part of the reason is that this is an area of ongoing research and development, and most of the tools and documentation are notably technical. Another reason is that these are not perfect tools. They can fail to show notable problems for models that are definitely misspecified, and flag models that are essentially okay. Sometimes they flag models that other indicators may suggest are better models relatively speaking, which actually isn’t a contradiction, but which may indicate an overfit situation. Another thing to think about from Gelman:</p>
<blockquote>
<p>The general issue is that with <em>unregularized</em> estimation such as least squares or maximum likelihood, adding parameters to a model (or making a model more complex) leads to overfitting. With regularized estimation such as multilevel modeling, Bayesian inference, lasso, deep learning, etc etc, the regularization adds complexity but in a way that reduces the problem of overfitting. So traditional notions of model complexity and tradeoffs are overturned. ~ Andrew Gelman</p>
</blockquote>
<p>So what should we do here? As an example, let’s start with our complex model and get the leave-one-out criterion measure. I will avoid as much technical jargon as possible so that the applied modeler can get on with things. The first part are the stats that are used in the previous model comparison and weighting, particularly the elpd_loo<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a>. We also get influence statistics for our observed data.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/brms/man/loo.brmsfit.html'>loo</a></span><span class='op'>(</span><span class='va'>model_baseline</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>
Computed from 1000 by 1000 log-likelihood matrix

         Estimate   SE
elpd_loo  -1500.5 21.9
p_loo        88.0  3.8
looic      3001.0 43.9
------
Monte Carlo SE of elpd_loo is 0.4.

Pareto k diagnostic values:
                         Count Pct.    Min. n_eff
(-Inf, 0.5]   (good)     992   99.2%   277       
 (0.5, 0.7]   (ok)         8    0.8%   176       
   (0.7, 1]   (bad)        0    0.0%   &lt;NA&gt;      
   (1, Inf)   (very bad)   0    0.0%   &lt;NA&gt;      

All Pareto k estimates are ok (k &lt; 0.7).
See help(&#39;pareto-k-diagnostic&#39;) for details.</code></pre>
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/brms/man/loo.brmsfit.html'>loo</a></span><span class='op'>(</span><span class='va'>model_interact_nonlin</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>
Computed from 1000 by 1000 log-likelihood matrix

         Estimate   SE
elpd_loo  -1454.0 22.1
p_loo        93.9  4.2
looic      2908.0 44.1
------
Monte Carlo SE of elpd_loo is 0.4.

Pareto k diagnostic values:
                         Count Pct.    Min. n_eff
(-Inf, 0.5]   (good)     995   99.5%   262       
 (0.5, 0.7]   (ok)         5    0.5%   102       
   (0.7, 1]   (bad)        0    0.0%   &lt;NA&gt;      
   (1, Inf)   (very bad)   0    0.0%   &lt;NA&gt;      

All Pareto k estimates are ok (k &lt; 0.7).
See help(&#39;pareto-k-diagnostic&#39;) for details.</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<ul>
<li>Resource: <a href="https://mc-stan.org/loo/reference/loo-glossary.html">The LOO glossary</a></li>
</ul>
<h4 id="pareto-values">Pareto values</h4>
<p>If you happen to see Pareto values in the ‘bad’ or ‘very bad’ group, what does it mean? You can read the definition provide <a href="https://mc-stan.org/loo/reference/loo-glossary.htmll#pareto-k-estimates">here</a>, but I doubt it will help many due to the background knowledge needed to parse it. However you can just understand it as an <em>outlier</em> diagnostic, and if it is a problem, it means your LOOIC may not be good for comparing models. As in the standard model setting setting, outliers indicate <em>model incompetence</em>, or in other words, its inability to understand such observations. Unless you have reason to suspect something wrong in the data (e.g. an incorrect value/typo), an outlier is a sign that your model is not able to capture the data fully. It definitely is not a reason to remove the observation.</p>
<p>If you have values &gt; .7, you may recalculate LOOIC with the options provided by the loo function, getting a better estimate that could then be used in, for example, model stacking for prediction. If you don’t have many discovered outliers, it probably won’t make much difference in your final estimates and conclusions. The output for Pareto values doesn’t even save the row identifying information that would make it easy to find which observations are the problem, but you can do something like the following.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='va'>problems</span> <span class='op'>=</span> <span class='va'>pareto_probs</span><span class='op'>$</span><span class='va'>pointwise</span> <span class='op'>%&gt;%</span> 
  <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>rowid_to_column</span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>group_by</span><span class='op'>(</span>flag <span class='op'>=</span> <span class='va'>influence_pareto_k</span> <span class='op'>&gt;</span> <span class='fl'>.5</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'><a href='https://rdrr.io/r/stats/filter.html'>filter</a></span><span class='op'>(</span><span class='va'>flag</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>pull</span><span class='op'>(</span><span class='va'>rowid</span><span class='op'>)</span>

<span class='va'>model_interact_nonlin</span><span class='op'>$</span><span class='va'>data</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>slice</span><span class='op'>(</span><span class='va'>problems</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>         y b1 b2         x1         x2          x3 group
1 4.835846  0  0  1.6595205 -0.1663858  0.03627905     6
2 5.976131  1  0  0.8606249 -0.6302693 -1.51768331     8
3 5.212302  1  0  2.0607612 -0.5475473 -1.38847085    58
4 2.139189  0  1 -0.8143745  0.5552497  2.07192410    69
5 3.181194  0  0 -1.6464138  0.5392800 -0.06286333    85</code></pre>
</div>
<p>As we might have expected, the observations with the more extreme target values are likely to be problems, but for some of these, there is nothing to suggest why they might be difficult, and it’s even harder to speculate in typical modeling situations with more predictors and complexity. Furthermore, outside of additional interactions, which might then become difficult to interpret, there is little we can do about this, or at least, is not obvious.</p>
<h4 id="the-rabbit-hole-of-model-comparsion">The rabbit hole of model comparsion</h4>
<p>If you start to look more into this, there are numerous technical articles, whole websites, and various discussions regarding how to go about these results. I’m guessing most don’t want to try and parse technical information only to still feel confused about what to actually do with the results. Furthermore, it’s not clear to me that issues regarding specific statistics for model comparison should be a basis for altering a model, unless there is an obvious path for doing so (which begs the question of why you weren’t already exploring that model). The suggestions typically amount to ‘Your model is misspecified’ no matter what the outcome. The are also suggestions that one should use posterior predictive checks (PPCs) are somehow to be used to detect the problem, but PPC don’t ‘detect’ anything in and of themselves, nor have any standard metric to report. , so this is all kind of a mess here.</p>
<p><a href="https://avehtari.github.io/modelselection/CV-FAQ.html#16_What_to_do_if_I_have_many_high_Pareto_(k)%E2%80%99s" class="uri">https://avehtari.github.io/modelselection/CV-FAQ.html#16_What_to_do_if_I_have_many_high_Pareto_(k)%E2%80%99s</a></p>
<p>The usual cases are</p>
<ul>
<li>misspecified models MC: all models are misspecified, so how will knowing this help us finish the project?</li>
<li>models with parameters which see the information only from one observation each (e.g. ‘random’ effect models) MC: sorry, but this is reality for many such models. Assuming we can’t increase this, which is practically every situation, what are we to do about it?</li>
<li>otherwise flexible models MC: how is ‘flexible’ defined? Overly vague priors?</li>
<li><a href="https://discourse.mc-stan.org/t/a-quick-note-what-i-infer-from-p-loo-and-pareto-k-values/3446">A quick note what I infer from p_loo and Pareto km values</a></li>
<li><a href="https://discourse.mc-stan.org/t/recommendations-for-what-to-do-when-k-exceeds-0-5-in-the-loo-package/3417">Recommendations for what to do when k exceeds 0.5 in the loo package?</a></li>
<li><a href="https://discourse.mc-stan.org/t/improve-model-with-some-observations-pareto-0-7/17500">Improve model with some observations Pareto &gt;0.7</a></li>
<li><a href="https://rawgit.com/avehtari/modelselection_tutorial/master/roaches.html#22_cross-validation_checking">Bayesian data analysis - roaches cross-validation demo</a></li>
<li><a href="https://avehtari.github.io/modelselection/CV-FAQ.html#16_What_to_do_if_I_have_many_high_Pareto_(k)%E2%80%99s">16 What to do if I have many high Pareto k’s?</a></li>
<li><a href="https://discourse.mc-stan.org/t/pareto-k-for-outlier-detection/12177/9">Pareto K for outlier detection 1</a></li>
</ul>
<blockquote>
<p>These measures are not independent. If there are many high Pareto k values as in case of model 4, then elpd_loo (or looic) can’t be trusted. Even if there would be no high Pareto k values, R^2 can’t be trusted if p_loo is relatively high compared to the total number of parameters or the number of observations as in case of model 2-4. So there is no contradiction here, but you need to take into account if diagnostics tell you that some other measures can’t be used. ~ <a href="https://discourse.mc-stan.org/t/good-pp-check-and-r-square-but-large-pareto-k-values/17678">Aki Vehtari</a></p>
</blockquote>
<h3 id="solutions-with-model-comparison">Solutions with Model Comparison</h3>
<ul>
<li>Avoid the problem and fit the model that includes everything of interest, assuming you have a decent data size to do so.</li>
<li>If some application performance is obvious and available to assess, pick a model that does best there.</li>
<li>If trying to select among many competing models, e.g. feature selection, you should consider why you are in this situation. If you don’t have much data, then the usual model selection criteria may lead you notably astray. If you have a lot of data, consider why you need to select a subset and not merely use all available. If you are somewhere in between, note that you’ll likely spend a lot more time here and still not be confident in the results. However, there are approaches, such as those in the <span class="pack" style="">projpred</span> package, that might be useful, but likely will only work for simpler models.</li>
<li>Use the metrics noted above, e.g. LOOIC,</li>
</ul>
<p><a href="https://avehtari.github.io/modelselection/CV-FAQ.html#5_How_to_use_cross-validation_for_model_selection" class="uri">https://avehtari.github.io/modelselection/CV-FAQ.html#5_How_to_use_cross-validation_for_model_selection</a></p>
<div class="layout-chunk" data-layout="l-body">

</div>
<h2 id="averaging-models">Averaging models</h2>
<p>With the previous statistics for model comparison we can obtain relative model weights, using the <span class="func" style="">model_weights</span> function. This essentially spreads the total probability of the models across all those being compared. These weights in turn allow us to obtain (weighted) average predictions. The key idea being that we do not select a ‘best’ model, but rather combine their results for predictive purposes<a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a>.</p>
<p>We can start by comparing the first two models. Adding the interactions helped, and comparing the weights suggests that it would be contributing most to averaged predictions.</p>
<div class="layout-chunk" data-layout="l-body">
<pre><code>Method: stacking
------
               weight
model_baseline 0.287 
model_interact 0.713 </code></pre>
</div>
<p>If we compare the baseline to our most complex model, almost the entirety of the weight is placed on the latter.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='fu'>loo_model_weights</span><span class='op'>(</span><span class='va'>model_baseline</span>, <span class='va'>model_interact_nonlin</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>Method: stacking
------
                      weight
model_baseline        0.000 
model_interact_nonlin 1.000 </code></pre>
</div>
<p>Now we compare all three, with roughly the same conclusion.</p>
<div class="layout-chunk" data-layout="l-body">
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:right;">
model_baseline
</th>
<th style="text-align:right;">
model_interact
</th>
<th style="text-align:right;">
model_interact_nonlin
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
1e-05
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.99999
</td>
</tr>
</tbody>
</table>
</div>
<p>Now what about those average predictions? Let’s create a data frame that sets the continuous covariates at their means, and at each level of the categorical covariates. We then will make average predictions for those observations using <span class="func" style="">pp_average</span>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='va'>prediction_data</span> <span class='op'>=</span> <span class='fu'>crossing</span><span class='op'>(</span>
  b1 <span class='op'>=</span> <span class='fl'>0</span><span class='op'>:</span><span class='fl'>1</span>,
  b2 <span class='op'>=</span> <span class='fl'>0</span><span class='op'>:</span><span class='fl'>1</span>,
  x1 <span class='op'>=</span> <span class='fl'>0</span>,
  x2 <span class='op'>=</span> <span class='fl'>0</span>,
  x3 <span class='op'>=</span> <span class='fl'>0</span>
<span class='op'>)</span>

<span class='va'>average_predictions</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/brms/man/pp_average.brmsfit.html'>pp_average</a></span><span class='op'>(</span>
  <span class='va'>model_baseline</span>,
  <span class='va'>model_interact</span>,
  <span class='va'>model_interact_nonlin</span>,
  newdata <span class='op'>=</span> <span class='va'>prediction_data</span>,
  re_formula <span class='op'>=</span> <span class='cn'>NA</span>
<span class='op'>)</span>
</code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:right;">
b1
</th>
<th style="text-align:right;">
b2
</th>
<th style="text-align:right;">
x1
</th>
<th style="text-align:right;">
x2
</th>
<th style="text-align:right;">
x3
</th>
<th style="text-align:right;">
Estimate
</th>
<th style="text-align:right;">
Est.Error
</th>
<th style="text-align:right;">
Q2.5
</th>
<th style="text-align:right;">
Q97.5
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2.712
</td>
<td style="text-align:right;">
1.002
</td>
<td style="text-align:right;">
0.711
</td>
<td style="text-align:right;">
4.632
</td>
</tr>
<tr>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2.354
</td>
<td style="text-align:right;">
0.996
</td>
<td style="text-align:right;">
0.460
</td>
<td style="text-align:right;">
4.330
</td>
</tr>
<tr>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2.973
</td>
<td style="text-align:right;">
1.008
</td>
<td style="text-align:right;">
1.047
</td>
<td style="text-align:right;">
5.053
</td>
</tr>
<tr>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
3.046
</td>
<td style="text-align:right;">
0.975
</td>
<td style="text-align:right;">
1.209
</td>
<td style="text-align:right;">
4.990
</td>
</tr>
</tbody>
</table>
</div>
<h2 id="cross-validation">Cross-validation</h2>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/HenrikBengtsson/future'>future</a></span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/future/man/plan.html'>plan</a></span><span class='op'>(</span><span class='va'>multiprocess</span><span class='op'>)</span>
<span class='va'>model_interact_nonlin_cv</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/brms/man/kfold.brmsfit.html'>kfold</a></span><span class='op'>(</span><span class='va'>model_interact_nonlin</span>, K <span class='op'>=</span> <span class='fl'>5</span>, chains <span class='op'>=</span> <span class='fl'>1</span>, save_fits <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/future/man/plan.html'>plan</a></span><span class='op'>(</span><span class='va'>sequential</span><span class='op'>)</span>

<span class='fu'><a href='https://rdrr.io/r/utils/str.html'>str</a></span><span class='op'>(</span><span class='va'>model_interact_nonlin_cv</span>, <span class='fl'>0</span><span class='op'>)</span>
<span class='va'>test</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/brms/man/kfold_predict.html'>kfold_predict</a></span><span class='op'>(</span><span class='va'>model_interact_nonlin_cv</span>, newdata <span class='op'>=</span> <span class='va'>prediction_data</span><span class='op'>)</span>

<span class='va'>rmse</span> <span class='op'>&lt;-</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>y</span>, <span class='va'>yrep</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='va'>yrep_mean</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colSums.html'>colMeans</a></span><span class='op'>(</span><span class='va'>yrep</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/base/MathFun.html'>sqrt</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/mean.html'>mean</a></span><span class='op'>(</span><span class='op'>(</span><span class='va'>yrep_mean</span> <span class='op'>-</span> <span class='va'>y</span><span class='op'>)</span><span class='op'>^</span><span class='fl'>2</span><span class='op'>)</span><span class='op'>)</span>
<span class='op'>}</span>
<span class='fu'>rmse</span><span class='op'>(</span>y <span class='op'>=</span> <span class='va'>test</span><span class='op'>$</span><span class='va'>y</span>, yrep <span class='op'>=</span> <span class='va'>test</span><span class='op'>$</span><span class='va'>yrep</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<h1 id="resources">Resources</h1>
<p><a href="https://statmodeling.stat.columbia.edu/2019/08/10/">Gelman’s prior check approach</a></p>
<p><a href="https://mc-stan.org/docs/2_24/stan-users-guide/problematic-posteriors-chapter.html" class="uri">https://mc-stan.org/docs/2_24/stan-users-guide/problematic-posteriors-chapter.html</a></p>
<p>Vehtari, A., Gelman, A., and Gabry, J. (2017a). Practical Bayesian model evaluation using leave-one-out cross-validation and WAIC. Statistics and Computing. 27(5), 1413–1432. <a href="doi:10.1007/s11222-016-9696-4" class="uri">doi:10.1007/s11222-016-9696-4</a> (<a href="http://link.springer.com/article/10.1007%2Fs11222-016-9696-4">journal version</a>, <a href="https://arxiv.org/abs/1507.04544">preprint arXiv:1507.04544</a>).</p>
<p>Yao, Y., Vehtari, A., Simpson, D., and Gelman, A. (2018) Using stacking to average Bayesian predictive distributions. Bayesian Analysis, advance publication, <a href="doi:10.1214/17-BA1091" class="uri">doi:10.1214/17-BA1091</a>. (<a href="https://projecteuclid.org/euclid.ba/1516093227">online</a>).</p>
<p>Vehtari, A., Simpson, D., Gelman, A., Yao, Y., and Gabry, J. (2019). Pareto smoothed importance sampling. <a href="https://arxiv.org/abs/1507.02646/">preprint arXiv:1507.02646</a></p>
<p>Gabry, J. , Simpson, D. , Vehtari, A. , Betancourt, M. and Gelman, A. (2019), Visualization in Bayesian workflow. J. R. Stat. Soc. A, 182: 389-402. <a href="doi:10.1111/rssa.12378" class="uri">doi:10.1111/rssa.12378</a>. (journal version, arXiv preprint, code on GitHub)</p>
<p><a href="https://discourse.mc-stan.org/t/a-quick-note-what-i-infer-from-p-loo-and-pareto-k-values/3446">Aki Vehtari’s A quick note what I infer from p_loo and Pareto k values</a></p>
<p><a href="https://github.com/stan-dev/stan/wiki/Prior-Choice-Recommendations" class="uri">https://github.com/stan-dev/stan/wiki/Prior-Choice-Recommendations</a></p>
<p><a href="https://mc-stan.org/docs/2_24/reference-manual/divergent-transitions.html" class="uri">https://mc-stan.org/docs/2_24/reference-manual/divergent-transitions.html</a></p>
<p><a href="https://discourse.mc-stan.org/t/divergent-transitions-a-primer/17099" class="uri">https://discourse.mc-stan.org/t/divergent-transitions-a-primer/17099</a></p>
<p><a href="https://mc-stan.org/docs/2_24/reference-manual/effective-sample-size-section.html" class="uri">https://mc-stan.org/docs/2_24/reference-manual/effective-sample-size-section.html</a></p>
<p><a href="https://mc-stan.org/bayesplot/articles/visual-mcmc-diagnostics.html" class="uri">https://mc-stan.org/bayesplot/articles/visual-mcmc-diagnostics.html</a></p>
<p><a href="http://mc-stan.org/cmdstanr/articles/cmdstanr.html">Use CmdStan to save memory</a></p>
<ul>
<li><a href="https://jrnold.github.io/bayesian_notes/">Jeffrey Arnold’s Bayesian Notes</a> has nice examples of many models and good summaries otherwise</li>
</ul>
<section class="footnotes" role="doc-endnotes">
<hr />
<ol>
<li id="fn1" role="doc-endnote"><p>In the same post, as a comment, Daniel Lakeland proposes an alternative approach is whether the posterior estimate falls within the 95% highest density interval of the prior. This is available via the method argument: <code>bayestestR::check_prior(model_baseline, method = 'lakeland')</code>.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2" role="doc-endnote"><p>Much like how AIC works with the -2*log likelihood.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3" role="doc-endnote"><p>Technically we can use WAIC to produce weights like we do with AIC, e.g. <code>exp(waic) / sum(exp(all_waics))</code>, but this isn’t recommended. The stacking approach allows similar models to share their weight, while more unique models will more or less keep their weight as additional models are added.<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4" role="doc-endnote"><p>The <code>looic</code> is just -2<em><code>elpd_loo</code>, as we often use -2</em>log likelihood (a.k.a. deviance) in standard approaches for AIC. In this case, <code>elpd_loo</code>, is a leave-one-out density. <code>p_loo</code> is the ‘effective number of parameters’, which users of penalized regression and mixed models will have some familiarity with.<a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn5" role="doc-endnote"><p>Some might be familiar with Bayesian model averaging. Conceptually we aren’t changing much, but BMA assumes that one of our models is the true model, while the stacking approach underlying these weights does not. It is also different from conventional stacking in machine learning in that we are trying to average posterior predictive distributions, rather than merely point estimates.<a href="#fnref5" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom"></div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
